From f299bdd208723809439ee92f60e6b8d366d4825d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Julian=20R=C3=BCth?= <julian.rueth@fsfe.org>
Date: Thu, 28 Dec 2023 01:37:49 -0600
Subject: [PATCH] Define TIMEVAL struct for Windows

we cannot include winsock2.h since that pulls in a min macro that is
incompatible with the one that GMP uses.
---
 source/libnormaliz/binomial.h  | 1 -
 source/libnormaliz/general.cpp | 4 +---
 source/libnormaliz/general.h   | 6 ++++++
 3 files changed, 7 insertions(+), 4 deletions(-)

diff --git a/source/libnormaliz/binomial.h b/source/libnormaliz/binomial.h
index 64d651e..299faf1 100644
--- a/source/libnormaliz/binomial.h
+++ b/source/libnormaliz/binomial.h
@@ -24,7 +24,6 @@
 #ifndef LATTICE_BINOMIAL_H
 #define LATTICE_BINOMIAL_H
 
-#include <sys/time.h>
 #include <string>
 #include "libnormaliz/matrix.h"
 
diff --git a/source/libnormaliz/general.cpp b/source/libnormaliz/general.cpp
index 6f578be..bf73895 100644
--- a/source/libnormaliz/general.cpp
+++ b/source/libnormaliz/general.cpp
@@ -25,9 +25,7 @@
 #include <csignal>
 #include "libnormaliz/general.h"
 
-#ifndef _MSC_VER
-#include <sys/time.h>
-#else
+#ifdef _MSC_VER
 #define WIN32_LEAN_AND_MEAN
 #include <Windows.h>
 #include <stdint.h> // portable: uint64_t   MSVC: __int64
diff --git a/source/libnormaliz/general.h b/source/libnormaliz/general.h
index b87624f..8ae1643 100644
--- a/source/libnormaliz/general.h
+++ b/source/libnormaliz/general.h
@@ -38,6 +38,12 @@
 #endif
 
 #ifdef _WIN32
+
+typedef struct timeval {
+  long tv_sec;
+  long tv_usec;
+};
+
 #if defined(DLL_EXPORT)
 #define NORMALIZ_DLL_EXPORT __declspec(dllexport)
 #elif defined(NORMALIZ_USE_DLL) && !defined(NORMALIZ_USE_STATIC)
-- 
2.43.0

